File: packages/framework/agents/specialized.ts
/**
 * JuliaOS Framework - Specialized Agents Module
 *
 * This module provides interfaces for specialized agent types in the JuliaOS framework.
 */

import { JuliaBridge } from '@juliaos/julia-bridge';
import { Agent, AgentType } from './index';
import { EventEmitter } from 'events';
import { AgentBlockchainIntegration } from './blockchain';
import { AgentMessaging } from './messaging';
import { AgentCollaboration } from './collaboration';

/**
 * Trading agent parameters
 */
export interface TradingAgentParameters {
  riskTolerance?: number;
  maxPositionSize?: number;
  takeProfit?: number;
  stopLoss?: number;
  rebalanceFrequency?: string;
  tradingHours?: {
    start: string;
    end: string;
    timezone: string;
  };
  allowedAssets?: string[];
  [key: string]: any;
}

/**
 * Trading strategy configuration
 */
export interface TradingStrategyConfig {
  name: string;
  parameters: Record<string, any>;
}

/**
 * Portfolio information
 */
export interface Portfolio {
  cash: Record<string, number>;
  assets: Record<string, number>;
  lastUpdated: string;
  [key: string]: any;
}

/**
 * Trade record
 */
export interface TradeRecord {
  id: string;
  timestamp: string;
  asset: string;
  amount: number;
  price: number;
  type: 'buy' | 'sell';
  value: number;
  status: string;
  [key: string]: any;
}

/**
 * Market analysis result
 */
export interface MarketAnalysis {
  timestamp: string;
  strategy: string;
  signals: Record<string, any>;
  marketData: Record<string, any>;
}

/**
 * TradingAgent class for interacting with trading agents
 */
export class TradingAgent extends EventEmitter {
  private bridge: JuliaBridge;
  private agentId: string;

  /**
   * Create a new TradingAgent
   *
   * @param bridge - JuliaBridge instance for communicating with the Julia backend
   * @param agentId - ID of the trading agent
   */
  constructor(bridge: JuliaBridge, agentId: string) {
    super();
    this.bridge = bridge;
    this.agentId = agentId;
  }

  /**
   * Initialize the trading agent
   *
   * @returns Promise with initialization result
   */
  async initialize(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('TradingAgent.initialize', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Execute a trade
   *
   * @param tradeData - Trade data
   * @returns Promise with trade execution result
   */
  async executeTrade(tradeData: {
    asset: string;
    amount: number;
    price: number;
    type: 'buy' | 'sell';
  }): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('TradingAgent.execute_trade', [this.agentId, tradeData]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Analyze market data
   *
   * @param marketData - Market data
   * @returns Promise with analysis result
   */
  async analyzeMarket(marketData: Record<string, any>): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('TradingAgent.analyze_market', [this.agentId, marketData]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Get portfolio information
   *
   * @returns Promise with portfolio information
   */
  async getPortfolio(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('TradingAgent.get_portfolio', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Set trading strategy
   *
   * @param strategy - Strategy name
   * @param config - Strategy configuration
   * @returns Promise with strategy setting result
   */
  async setStrategy(strategy: string, config: Record<string, any> = {}): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('TradingAgent.set_strategy', [this.agentId, strategy, config]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }
}

/**
 * Monitor agent parameters
 */
export interface MonitorAgentParameters {
  checkInterval?: number;
  alertChannels?: string[];
  maxAlertsPerHour?: number;
  alertCooldown?: number;
  dataSources?: string[];
  [key: string]: any;
}

/**
 * Alert configuration
 */
export interface AlertConfig {
  asset: string;
  conditionType: string;
  condition: string;
  threshold: number;
  message: string;
  enabled?: boolean;
  channels?: string[];
  cooldown?: number;
}

/**
 * Alert record
 */
export interface AlertRecord {
  id: string;
  asset: string;
  conditionType: string;
  condition: string;
  threshold: number;
  message: string;
  triggeredAt: string;
  assetData: Record<string, any>;
}

/**
 * MonitorAgent class for interacting with monitoring agents
 */
export class MonitorAgent extends EventEmitter {
  private bridge: JuliaBridge;
  private agentId: string;

  /**
   * Create a new MonitorAgent
   *
   * @param bridge - JuliaBridge instance for communicating with the Julia backend
   * @param agentId - ID of the monitor agent
   */
  constructor(bridge: JuliaBridge, agentId: string) {
    super();
    this.bridge = bridge;
    this.agentId = agentId;
  }

  /**
   * Initialize the monitor agent
   *
   * @returns Promise with initialization result
   */
  async initialize(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('MonitorAgent.initialize', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Configure alerts
   *
   * @param alertConfigs - Alert configurations
   * @returns Promise with configuration result
   */
  async configureAlerts(alertConfigs: AlertConfig[]): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('MonitorAgent.configure_alerts', [this.agentId, alertConfigs]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Check alert conditions against market data
   *
   * @param marketData - Market data
   * @returns Promise with check result
   */
  async checkConditions(marketData: Record<string, any>): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('MonitorAgent.check_conditions', [this.agentId, marketData]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Get active alerts
   *
   * @returns Promise with active alerts
   */
  async getAlerts(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('MonitorAgent.get_alerts', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Get alert history
   *
   * @returns Promise with alert history
   */
  async getAlertHistory(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('MonitorAgent.get_alert_history', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }
}

/**
 * Arbitrage agent parameters
 */
export interface ArbitrageAgentParameters {
  minProfitThreshold?: number;
  maxPositionSize?: number;
  gasCostBuffer?: number;
  executionTimeout?: number;
  maxSlippage?: number;
  chains?: string[];
  exchanges?: string[];
  assets?: string[];
  scanInterval?: number;
  riskLevel?: 'low' | 'medium' | 'high';
  [key: string]: any;
}

/**
 * Arbitrage opportunity
 */
export interface ArbitrageOpportunity {
  id: string;
  type: string;
  asset?: string;
  sourceExchange?: string;
  targetExchange?: string;
  sourceChain?: string;
  targetChain?: string;
  sourcePrice?: number;
  targetPrice?: number;
  expectedProfitPct: number;
  expectedProfit: number;
  amount: number;
  timestamp: string;
  [key: string]: any;
}

/**
 * Arbitrage execution
 */
export interface ArbitrageExecution {
  id: string;
  opportunityId: string;
  opportunity: ArbitrageOpportunity;
  startTime: string;
  endTime?: string;
  status: string;
  actualProfit?: number;
  actualProfitPct?: number;
  transactions: Record<string, any>[];
  error?: string;
}

/**
 * ArbitrageAgent class for interacting with arbitrage agents
 */
export class ArbitrageAgent extends EventEmitter {
  private bridge: JuliaBridge;
  private agentId: string;

  /**
   * Create a new ArbitrageAgent
   *
   * @param bridge - JuliaBridge instance for communicating with the Julia backend
   * @param agentId - ID of the arbitrage agent
   */
  constructor(bridge: JuliaBridge, agentId: string) {
    super();
    this.bridge = bridge;
    this.agentId = agentId;
  }

  /**
   * Initialize the arbitrage agent
   *
   * @returns Promise with initialization result
   */
  async initialize(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('ArbitrageAgent.initialize', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Find arbitrage opportunities
   *
   * @param marketData - Market data
   * @returns Promise with opportunities found
   */
  async findOpportunities(marketData: Record<string, any>): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('ArbitrageAgent.find_opportunities', [this.agentId, marketData]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Execute an arbitrage opportunity
   *
   * @param opportunityId - Opportunity ID
   * @returns Promise with execution result
   */
  async executeArbitrage(opportunityId: string): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('ArbitrageAgent.execute_arbitrage', [this.agentId, opportunityId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Get arbitrage history
   *
   * @returns Promise with arbitrage history
   */
  async getHistory(): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('ArbitrageAgent.get_history', [this.agentId]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Set arbitrage parameters
   *
   * @param parameters - Arbitrage parameters
   * @returns Promise with parameter setting result
   */
  async setParameters(parameters: Partial<ArbitrageAgentParameters>): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('ArbitrageAgent.set_parameters', [this.agentId, parameters]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }
}

/**
 * LLM configuration
 */
export interface LLMConfig {
  provider: string;
  apiKey?: string;
  model?: string;
  temperature?: number;
  maxTokens?: number;
  topP?: number;
  frequencyPenalty?: number;
  presencePenalty?: number;
  timeout?: number;
  retryCount?: number;
  retryDelay?: number;
}

/**
 * LLMIntegration class for interacting with LLM providers
 */
export class LLMIntegration extends EventEmitter {
  private bridge: JuliaBridge;

  /**
   * Create a new LLMIntegration
   *
   * @param bridge - JuliaBridge instance for communicating with the Julia backend
   */
  constructor(bridge: JuliaBridge) {
    super();
    this.bridge = bridge;
  }

  /**
   * Initialize the LLM integration
   *
   * @param config - LLM configuration
   * @returns Promise with initialization result
   */
  async initialize(config: LLMConfig): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('LLMIntegration.initialize_llm', [config]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Generate a response from the LLM
   *
   * @param prompt - Prompt to send to the LLM
   * @param config - Optional configuration overrides
   * @returns Promise with LLM response
   */
  async generateResponse(prompt: string, config: Partial<LLMConfig> = {}): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('LLMIntegration.generate_response', [prompt, config]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Generate a structured output from the LLM
   *
   * @param prompt - Prompt to send to the LLM
   * @param outputSchema - Schema defining the expected output structure
   * @param config - Optional configuration overrides
   * @returns Promise with structured LLM response
   */
  async generateStructuredOutput(
    prompt: string,
    outputSchema: Record<string, any>,
    config: Partial<LLMConfig> = {}
  ): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('LLMIntegration.generate_structured_output', [
        prompt,
        outputSchema,
        config
      ]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Analyze text using the LLM
   *
   * @param text - Text to analyze
   * @param analysisType - Type of analysis to perform
   * @param config - Optional configuration overrides
   * @returns Promise with analysis result
   */
  async analyzeText(
    text: string,
    analysisType: 'sentiment' | 'entities' | 'summary' | 'keywords',
    config: Partial<LLMConfig> = {}
  ): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('LLMIntegration.analyze_text', [text, analysisType, config]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Generate an embedding vector for text
   *
   * @param text - Text to embed
   * @param config - Optional configuration overrides
   * @returns Promise with embedding result
   */
  async generateEmbedding(text: string, config: Partial<LLMConfig> = {}): Promise<Record<string, any>> {
    try {
      return await this.bridge.execute('LLMIntegration.generate_embedding', [text, config]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Create a prompt from a template
   *
   * @param templateName - Name of the template
   * @param variables - Variables to fill in the template
   * @returns Promise with the filled prompt
   */
  async createPrompt(templateName: string, variables: Record<string, string>): Promise<string> {
    try {
      return await this.bridge.execute('LLMIntegration.create_prompt', [templateName, variables]);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }

  /**
   * Get the list of supported LLM providers
   *
   * @returns Promise with the list of providers
   */
  async getProviders(): Promise<string[]> {
    try {
      return await this.bridge.execute('LLMIntegration.get_providers', []);
    } catch (error) {
      this.emit('error', error);
      throw error;
    }
  }
}

/**
 * Factory function to create a specialized agent based on type
 *
 * @param bridge - JuliaBridge instance
 * @param agent - Agent object
 * @returns Specialized agent instance
 */
export function createSpecializedAgent(bridge: JuliaBridge, agent: Agent): TradingAgent | MonitorAgent | ArbitrageAgent {
  switch (agent.type) {
    case AgentType.TRADING:
      return new TradingAgent(bridge, agent.id);
    case AgentType.MONITOR:
      return new MonitorAgent(bridge, agent.id);
    case AgentType.ARBITRAGE:
      return new ArbitrageAgent(bridge, agent.id);
    default:
      throw new Error(`No specialized implementation for agent type: ${agent.type}`);
  }
}

/**
 * Create a blockchain integration for an agent
 *
 * @param bridge - JuliaBridge instance
 * @param agentId - Agent ID
 * @returns AgentBlockchainIntegration instance
 */
export function createAgentBlockchainIntegration(bridge: JuliaBridge, agentId: string): AgentBlockchainIntegration {
  return new AgentBlockchainIntegration(bridge, agentId);
}

/**
 * Create a messaging system for an agent
 *
 * @param bridge - JuliaBridge instance
 * @param agentId - Agent ID
 * @returns AgentMessaging instance
 */
export function createAgentMessaging(bridge: JuliaBridge, agentId: string): AgentMessaging {
  return new AgentMessaging(bridge, agentId);
}

/**
 * Create a collaboration system for an agent
 *
 * @param bridge - JuliaBridge instance
 * @param agentId - Agent ID
 * @returns AgentCollaboration instance
 */
export function createAgentCollaboration(bridge: JuliaBridge, agentId: string): AgentCollaboration {
  return new AgentCollaboration(bridge, agentId);
}

File: packages/python-wrapper/juliaos/agents/specialized.py
"""
Specialized agent classes for the JuliaOS Python wrapper.
"""

from typing import Dict, Any, List, Optional, Union

from ..bridge import JuliaBridge
from ..exceptions import AgentError
from .agent import Agent


class TradingAgent(Agent):
    """
    Trading agent class.
    
    This class provides methods for interacting with a trading agent in the JuliaOS Framework.
    """
    
    async def initialize(self) -> Dict[str, Any]:
        """
        Initialize the trading agent.
        
        Returns:
            Dict[str, Any]: Initialization result
        
        Raises:
            AgentError: If initialization fails
        """
        try:
            result = await self.bridge.execute("TradingAgent.initialize", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to initialize trading agent: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error initializing trading agent: {e}")
            raise
    
    async def execute_trade(self, trade_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute a trade.
        
        Args:
            trade_data: Trade data
        
        Returns:
            Dict[str, Any]: Trade execution result
        
        Raises:
            AgentError: If trade execution fails
        """
        try:
            result = await self.bridge.execute("TradingAgent.execute_trade", [self.id, trade_data])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to execute trade: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error executing trade: {e}")
            raise
    
    async def analyze_market(self, market_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Analyze market data.
        
        Args:
            market_data: Market data
        
        Returns:
            Dict[str, Any]: Analysis result
        
        Raises:
            AgentError: If market analysis fails
        """
        try:
            result = await self.bridge.execute("TradingAgent.analyze_market", [self.id, market_data])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to analyze market: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error analyzing market: {e}")
            raise
    
    async def get_portfolio(self) -> Dict[str, Any]:
        """
        Get portfolio information.
        
        Returns:
            Dict[str, Any]: Portfolio information
        
        Raises:
            AgentError: If portfolio retrieval fails
        """
        try:
            result = await self.bridge.execute("TradingAgent.get_portfolio", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to get portfolio: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error retrieving portfolio: {e}")
            raise
    
    async def set_strategy(self, strategy: str, config: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Set trading strategy.
        
        Args:
            strategy: Strategy name
            config: Strategy configuration
        
        Returns:
            Dict[str, Any]: Strategy setting result
        
        Raises:
            AgentError: If strategy setting fails
        """
        try:
            result = await self.bridge.execute("TradingAgent.set_strategy", [
                self.id,
                strategy,
                config or {}
            ])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to set strategy: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error setting strategy: {e}")
            raise


class MonitorAgent(Agent):
    """
    Monitor agent class.
    
    This class provides methods for interacting with a monitor agent in the JuliaOS Framework.
    """
    
    async def initialize(self) -> Dict[str, Any]:
        """
        Initialize the monitor agent.
        
        Returns:
            Dict[str, Any]: Initialization result
        
        Raises:
            AgentError: If initialization fails
        """
        try:
            result = await self.bridge.execute("MonitorAgent.initialize", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to initialize monitor agent: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error initializing monitor agent: {e}")
            raise
    
    async def configure_alerts(self, alert_configs: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Configure alerts.
        
        Args:
            alert_configs: Alert configurations
        
        Returns:
            Dict[str, Any]: Alert configuration result
        
        Raises:
            AgentError: If alert configuration fails
        """
        try:
            result = await self.bridge.execute("MonitorAgent.configure_alerts", [self.id, alert_configs])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to configure alerts: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error configuring alerts: {e}")
            raise
    
    async def check_conditions(self, market_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Check alert conditions against market data.
        
        Args:
            market_data: Market data
        
        Returns:
            Dict[str, Any]: Check result
        
        Raises:
            AgentError: If condition checking fails
        """
        try:
            result = await self.bridge.execute("MonitorAgent.check_conditions", [self.id, market_data])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to check conditions: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error checking conditions: {e}")
            raise
    
    async def get_alerts(self) -> Dict[str, Any]:
        """
        Get active alerts.
        
        Returns:
            Dict[str, Any]: Active alerts
        
        Raises:
            AgentError: If alert retrieval fails
        """
        try:
            result = await self.bridge.execute("MonitorAgent.get_alerts", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to get alerts: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error retrieving alerts: {e}")
            raise
    
    async def get_alert_history(self) -> Dict[str, Any]:
        """
        Get alert history.
        
        Returns:
            Dict[str, Any]: Alert history
        
        Raises:
            AgentError: If alert history retrieval fails
        """
        try:
            result = await self.bridge.execute("MonitorAgent.get_alert_history", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to get alert history: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error retrieving alert history: {e}")
            raise


class ArbitrageAgent(Agent):
    """
    Arbitrage agent class.
    
    This class provides methods for interacting with an arbitrage agent in the JuliaOS Framework.
    """
    
    async def initialize(self) -> Dict[str, Any]:
        """
        Initialize the arbitrage agent.
        
        Returns:
            Dict[str, Any]: Initialization result
        
        Raises:
            AgentError: If initialization fails
        """
        try:
            result = await self.bridge.execute("ArbitrageAgent.initialize", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to initialize arbitrage agent: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error initializing arbitrage agent: {e}")
            raise
    
    async def find_opportunities(self, market_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Find arbitrage opportunities.
        
        Args:
            market_data: Market data
        
        Returns:
            Dict[str, Any]: Opportunities found
        
        Raises:
            AgentError: If opportunity finding fails
        """
        try:
            result = await self.bridge.execute("ArbitrageAgent.find_opportunities", [self.id, market_data])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to find opportunities: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error finding opportunities: {e}")
            raise
    
    async def execute_arbitrage(self, opportunity_id: str) -> Dict[str, Any]:
        """
        Execute an arbitrage opportunity.
        
        Args:
            opportunity_id: Opportunity ID
        
        Returns:
            Dict[str, Any]: Execution result
        
        Raises:
            AgentError: If arbitrage execution fails
        """
        try:
            result = await self.bridge.execute("ArbitrageAgent.execute_arbitrage", [self.id, opportunity_id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to execute arbitrage: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error executing arbitrage: {e}")
            raise
    
    async def get_history(self) -> Dict[str, Any]:
        """
        Get arbitrage history.
        
        Returns:
            Dict[str, Any]: Arbitrage history
        
        Raises:
            AgentError: If history retrieval fails
        """
        try:
            result = await self.bridge.execute("ArbitrageAgent.get_history", [self.id])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to get history: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error retrieving history: {e}")
            raise
    
    async def set_parameters(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """
        Set arbitrage parameters.
        
        Args:
            parameters: Arbitrage parameters
        
        Returns:
            Dict[str, Any]: Parameter setting result
        
        Raises:
            AgentError: If parameter setting fails
        """
        try:
            result = await self.bridge.execute("ArbitrageAgent.set_parameters", [self.id, parameters])
            
            if not result.get("success", False):
                raise AgentError(f"Failed to set parameters: {result.get('error', 'Unknown error')}")
            
            return result
        except Exception as e:
            if not isinstance(e, AgentError):
                raise AgentError(f"Error setting parameters: {e}")
            raise
